### PR TITLE
EMTF NNv16 Implementation

### PR DESCRIPTION
**Pull request by eyigitba**
#### PR description:

This PR reworks the EMTF displaced pT assignment blocks to implement NNv16. Any other version of NN before this was merely experimental on software side and never worked at P5. After talking with @perrotta , I decided to remove the support for previous versions completely since they should never be used.

- Most of the rework is in `PtAssignmentEngineDxy.cc` to implement the new input/output for the NNv16. 
- `Common.h` changes since the NN input features have different size now
- `PtAssignment.cc` change due to NN output now being pT instead of 1/pT
- `PtAssignmentEngineAux.cc` change due to new `dxy` binning in firmware
- `simEmtfDigis.py` change for the new NN version.

This PR requires https://github.com/cms-data/L1Trigger-L1TMuon/pull/24 for tests.

This PR will be backported to 13_1_X and 13_0_X to make emulator agree with online firmware. 

#### PR validation:

Validated by checking firmware/emulator comparisons. Also ran `runTheMatrix.py -l limited -i all --ibeos` no failures (other than a fail due to input file) were seen.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

FYI @caruta @elfontan

### CODE REVIEW

**Comment by perrotta** [comment]
test parameters:
- pull_request = https://github.com/cms-data/L1Trigger-L1TMuon/pull/24

**Comment by perrotta** [comment]
please test

File: L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc [hunk 1]
```diff
@@ -61,38 +61,26 @@ void PtAssignmentEngineDxy::preprocessing_dxy(const EMTFTrack& track, emtf::Feat
...
-  int dTh_12, dTh_13, dTh_14, dTh_23, dTh_24, dTh_34;
-  int fr_1;
-  int bend_1, bend_2, bend_3, bend_4;
-  int rpc_1, rpc_2, rpc_3, rpc_4;
-  int St1_ring2 = data.st1_ring2;
-
-  int pat1 = -99, pat2 = -99, pat3 = -99, pat4 = -99;
-
   // // Which stations have hits
   int st1 = (track.Mode() >= 8);
```
Comment (perrotta) [inline L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc:85]: Why not `bool`, as they are only used as such?

File: L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc [hunk 2]
```diff
@@ -101,140 +89,93 @@ void PtAssignmentEngineDxy::preprocessing_dxy(const EMTFTrack& track, emtf::Feat
...
+    x_csc_pattern[3] = data.cpattern[3];
 
   // Calculate delta phi
-  dPhi_12 = (data.delta_ph[0] != invalid_dphi) ? data.delta_ph[0] * (data.sign_ph[0] ? 1 : -1) : 0;
-  dPhi_13 = (data.delta_ph[1] != invalid_dphi) ? data.delta_ph[1] * (data.sign_ph[1] ? 1 : -1) : 0;
-  dPhi_14 = (data.delta_ph[2] != invalid_dphi) ? data.delta_ph[2] * (data.sign_ph[2] ? 1 : -1) : 0;
-  dPhi_23 = (data.delta_ph[3] != invalid_dphi) ? data.delta_ph[3] * (data.sign_ph[3] ? 1 : -1) : 0;
-  dPhi_24 = (data.delta_ph[4] != invalid_dphi) ? data.delta_ph[4] * (data.sign_ph[4] ? 1 : -1) : 0;
-  dPhi_34 = (data.delta_ph[5] != invalid_dphi) ? data.delta_ph[5] * (data.sign_ph[5] ? 1 : -1) : 0;
+  x_dphi[0] = (data.delta_ph[0] != invalid_dphi) ? data.delta_ph[0] : 0;
```
Comment (perrotta) [inline L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc:101]: A `for` loop would simplify the code now that you have indexed also the name of the variables

File: L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc [hunk 3]
```diff
@@ -61,38 +61,26 @@ void PtAssignmentEngineDxy::preprocessing_dxy(const EMTFTrack& track, emtf::Feat
...
-  std::array<float, 1> x_me11ring;
-  std::array<float, 4> x_rpcbit;
+  std::array<float, 4> x_csc_pattern;
 
   // Initialize to zeros
   x_dphi.fill(0);
+  x_dphi_sign.fill(0);
   x_dtheta.fill(0);
+  x_dtheta_sign.fill(0);
   //
```
Comment (perrotta) [inline L1Trigger/L1TMuonEndCap/src/PtAssignmentEngineDxy.cc:75]: These are not needed, as they are filled in any case
```suggestion
```

**Comment by aloeliger** [comment]
@eyigitba Any news from your side on Andrea's suggestions?

**Comment by eyigitba** [comment]
@aloeliger @perrotta I returned from vacation yesterday. I'll implement the suggestions soon (hopefully this week)

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
+l1

**Comment by perrotta** [comment]
+1

**Pull request merged**